File: src/main/java/ru/vmsystems/template/domain/service/MaterialService.java
package ru.vmsystems.template.domain.service;

import org.dozer.DozerBeanMapper;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import ru.vmsystems.template.domain.model.MaterialEntity;
import ru.vmsystems.template.infrastructure.persistence.MaterialRepository;
import ru.vmsystems.template.interfaces.dto.MaterialDto;

import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;

@Service
public class MaterialService extends BackService {

    @Autowired
    private MaterialRepository repository;
    @Autowired
    private DozerBeanMapper mapper;

    public List<MaterialDto> get() {
        return repository.getByCompanyId(getCompanyId()).stream()
                .map(entity -> mapper.map(entity, MaterialDto.class))
                .collect(Collectors.toList());
    }

    public Optional<MaterialDto> get(@NotNull Long id) {
        MaterialEntity entity = repository.findOne(id);
        if (entity == null) {
            return Optional.empty();
        }

        return Optional.of(mapper.map(entity, MaterialDto.class));
    }

    @NotNull
    public MaterialDto create(@NotNull MaterialDto dto) {
        return update(null, dto);
    }

    @NotNull
    public MaterialDto update(@Nullable Long id, @NotNull MaterialDto dto) {

        dto.setId(id);
        MaterialEntity entity = mapper.map(dto, MaterialEntity.class);
        entity.setCompany(getCompany());
        entity = repository.save(entity);

        return mapper.map(entity, MaterialDto.class);
    }

    public void delete(Long orderId) {
        repository.delete(orderId);
    }
}

File: src/main/java/ru/vmsystems/glasscloud/domain/material/MaterialService.kt
package ru.vmsystems.glasscloud.domain.material

import org.springframework.stereotype.Service
import ru.vmsystems.glasscloud.domain.BackService
import ru.vmsystems.glasscloud.domain.SessionService
import ru.vmsystems.glasscloud.domain.material.color.transform
import ru.vmsystems.glasscloud.domain.material.type.transform
import ru.vmsystems.glasscloud.handler.BusinessException
import java.util.*

@Service
class MaterialService(
        private val repository: MaterialRepository,
        private val sessionService: SessionService
) : BackService() {

    fun getAll(): List<MaterialDto> {
        return repository.getByCompanyId(sessionService.currentCompanyId)
                .map { it.transform() }
                .toList()
    }

    fun get(id: UUID): MaterialDto {
        val entity = repository.getById(id) ?: throw BusinessException("Не найден тип материала")
        return entity.transform()
    }

    fun save(dto: MaterialDto): MaterialDto {
        val entity = dto.transform(sessionService.currentCompanyId)
        val result = repository.save(entity)
        return result.transform()
    }

    fun delete(orderId: UUID) {
        repository.deleteById(orderId)
    }
}

fun MaterialDto.transform(currentCompanyId: UUID? = null): MaterialEntity {
    return MaterialEntity(
            id = id,
            deleted = deleted,
            depth = depth,
            length = length,
            width = width,
            price = price,
            description = description,
            companyId = currentCompanyId ?: companyId!!,
            type = type.transform(),
            color = color.transform()
    )
}

fun MaterialEntity.transform(): MaterialDto {
    return MaterialDto(
            id = id,
            deleted = deleted,
            depth = depth,
            length = length,
            width = width,
            price = price,
            description = description,
            companyId = companyId,
            type = type.transform(),
            color = color.transform()
    )
}
